### PR TITLE
Added frame selection support.

### PR DESCRIPTION
**Pull request by terrylucas**
Code to support selection of bars in the frame chart.
Cleaned up how traces are created and the list order to eliminate some brittleness.

### CODE REVIEW

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart [hunk 1]
```diff
@@ -97,28 +97,78 @@ class PlotlyDivGraph extends CoreElement {
...
-      final TimelineFrame timelineFrame = _frames[xPosition];
-      framesBarChart.setSelected(timelineFrame);
+
+    final List<int> pointNumbers = [];
+    final List<int> xValues = [];
+    final List<num> yValues = [];
+    final List<int> traces = [];
+
+    for (Point pt in data.points) {
+      // Don't allow selecting an already selected bar.
```
Comment (jacob314) [inline packages/devtools/lib/src/timeline/frames_bar_chart.dart]: what happens when an already selected bar is selected? 
Should we bail out of the click logic early if the selection didn't change or is there a reason we can't?

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart [hunk 2]
```diff
@@ -97,28 +97,78 @@ class PlotlyDivGraph extends CoreElement {
...
+        xValues.add(pt.x);
+        yValues.add(pt.y);
+      }
+    }
+
+    // Create selection once.
+    currentSelection ??= Selection(frameGraph, element);
+
+    // If this bar isn't currently selected then select the bar clicked.
+    if (xValues.length == 2 &&
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_chart.dart]: Why xValues.length == 2? Can you add a comment explaining this.

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart [hunk 3]
```diff
@@ -97,28 +97,78 @@ class PlotlyDivGraph extends CoreElement {
...
+            (pt.curveNumber == FramesBarPlotly.gpuSelectTraceIndex) ||
+                (pt.curveNumber == FramesBarPlotly.cpuSelectTraceIndex);
         hoverDisplay.add(
             HoverFX(curveNumber: pt.curveNumber, pointNumber: pt.pointNumber));
       }
     }
 
+    if (selectionTrace) {
+      // Hide the hover of the gpu good/jank & cpu good/jank trace of the
+      // selected bar.
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_chart.dart]: is there a reason we don't still want to show the duration values on hover?

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 4]
```diff
@@ -109,37 +163,23 @@ class FramesBarPlotly {
         color: colorToCss(cpuJankColor),
       ),
       width: [0],
-    );
+    ));
 
-    final Data traceGpuJank = Data(
+    // trace CPU Seleect
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: typo in select

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 5]
```diff
@@ -59,39 +67,85 @@ class FramesBarPlotly {
...
-      name: 'CPU',
+      name: 'GPU',
       hoverinfo: 'y+name',
       marker: Marker(
-        color: colorToCss(mainCpuColor),
+        color: colorToCss(mainGpuColor),
       ),
       width: [0],
-    );
+    ));
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: can we add a trailing comma after the Data() object for all these `allTraces.insert(...)` calls so that we are using fluter style?

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 6]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+      selectInfo[0].xValue == xValues[0] &&
+      selectInfo[1].xValue == xValues[1];
+
+  int get selectedPointNumber =>
+      selectInfo.isNotEmpty ? selectInfo[0].ptNumber : -1;
+
+  void select(List<int> pointNumbers, List<int> traces, List<int> xValues,
+      List<num> yValues) {
+    // Supports one bar selection and not selecting a currently selected bar.
+    assert(traces.length == 2 &&
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: perhaps make 2 a const so that we know what it means and can use it here and elsewhere

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 7]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+    // Construct our new selection info.
+    int index = 0;
+    for (var trace in traces) {
+      selectInfo.add(SelectTrace(
+          trace, pointNumbers[index], xValues[index], yValues[index]));
+      index++;
+    }
+
+    // Make room for our selection bar remove the data we're selecting it will
+    // exist in the selection traces.
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: can you rephrase this comment or split into two sentences? I'm not sure what you mean here.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 8]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+  SelectTrace(this.traceIndex, this.ptNumber, this.xValue, this.yValue);
+
+  final int traceIndex;
+  final int ptNumber;
+  final int xValue;
+  final num yValue;
+}
+
+class Selection {
+  Selection(this._domName, dynamic graphDiv)
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: add trailing comma after `graphDiv`

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 9]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+  bool isSelected(List<int> xValues) =>
+      selectInfo.length == 2 &&
+      selectInfo[0].xValue == xValues[0] &&
+      selectInfo[1].xValue == xValues[1];
+
+  int get selectedPointNumber =>
+      selectInfo.isNotEmpty ? selectInfo[0].ptNumber : -1;
+
+  void select(List<int> pointNumbers, List<int> traces, List<int> xValues,
+      List<num> yValues) {
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: add trailing comma after yValues

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 10]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+      if (oldTrace1 == traces[1] && pointNumbers[1] >= oldPtNum1) {
+        pointNumbers[1] += 1;
+      }
+    }
+
+    // Construct our new selection info.
+    int index = 0;
+    for (var trace in traces) {
+      selectInfo.add(SelectTrace(
+          trace, pointNumbers[index], xValues[index], yValues[index]));
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: add trailing comma after `yValues[index]`

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 11]
```diff
@@ -305,3 +345,136 @@ class FramesBarPlotly {
...
+
+    // Construct the hover names for each selection trace.
+    final String gpuSelectionHoverName =
+        traces[0] == FramesBarPlotly.gpuGoodTraceIndex ? 'GPU' : 'GPU Jank';
+    final String cpuSelectionHoverName =
+        traces[1] == FramesBarPlotly.cpuGoodTraceIndex ? 'CPU' : 'CPU Jank';
+
+    // Update the hovers for the selection traces.
+    Plotly.restyle(_domName, 'name', [gpuSelectionHoverName],
+        [FramesBarPlotly.gpuSelectTraceIndex]);
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: nit: add trailing comma here and below

File: packages/devtools/web/index.html [hunk 12]
```diff
@@ -85,6 +85,20 @@
...
+            var data = {};
+            var xData = [];
+            var yData = [];
+            var traces = [];
+
+            _copyTrace(x0, y0, orgTraces[0], xData, yData, traces);
+            _copyTrace(x1, y1, orgTraces[1], xData, yData, traces);
+
+            data = { x: xData, y: yData};
+
```
Comment (kenzieschmoll) [inline packages/devtools/web/index.html]: nit: remove extra space before x or after yData (not sure what the proper style should be here but it should at least be consistent)

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart (hunk 1 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_chart.dart]: This comment was in the wrong place.  However, there is a quick bailout  !currentSelection.isSelected(newSelection)) I can't bailout until I have the SelectTrace constructed.  See:

    // Don't allow selecting an already selected bar.  If this bar isn't
    // currently selected then select the bar clicked.
    if (newSelection.length == 2 &&
        !currentSelection.isSelected(newSelection)) {
      currentSelection.select(newSelection);

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart (hunk 2 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_chart.dart]: It been changed to 
`newSelection.length
`
but two traces should exist, a point is always in 2 traces and only 2, otherwise there's a problem.  I'll add a comment it was strange before I think it's better now.

File: packages/devtools/lib/src/timeline/frames_bar_chart.dart (hunk 3 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_chart.dart]: There was before I was removing the point from a trace.  As I cleaned up the SelectTrace in plotlyClick function I removed this too.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 4 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Done.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 5 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Done.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 6 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: It means two traces per data point.  I'll add a constant.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 8 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: done

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 9 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Signature has changed just one parameter now.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 10 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Code has been removed SelectTrace is constructed elsewhere but it been fixed there as well.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 11 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Done.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart [hunk 13]
```diff
@@ -59,87 +67,131 @@ class FramesBarPlotly {
...
+          x: [xCoordNotUsed],
+          type: 'bar',
+          legendgroup: 'good_group',
+          name: 'GPU',
+          hoverinfo: 'y+name',
+          marker: Marker(
+            color: colorToCss(mainGpuColor),
+          ),
+          width: [0],
+        ));
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: can you add the trailing comma between the final two parentheses? that will bring the final paren and the semicolon down to be vertically inline with `allTraces...`. If you add the trailing comma between the final two parens and then run dartfmt it should fix it automatically.

Same for all cases below.

File: packages/devtools/web/index.html (hunk 12 above)
Comment (terrylucas) [reply packages/devtools/web/index.html]: Done removed the extra space.

File: packages/devtools/lib/src/timeline/frames_bar_plotly.dart (hunk 13 above)
Comment (terrylucas) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: Yep already did that when you made the earlier suggestion.

Comment (kenzieschmoll) [reply packages/devtools/lib/src/timeline/frames_bar_plotly.dart]: sorry I thought you had already pushed that commit. Nevermind :)
